Add client-side session caching

tmshort opened the pull request:
Add support for client-side caching.
Uses an application defined "cache_id" to identify a connection. This cache_id may be an IP address, DNS name or anything else the client wishes to use.

Replacement of #5932 

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [X] documentation is added or updated
- [X] tests are added or updated

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 278 days ago

tmshort committed:
Add client-side session caching

Add support for client-side caching.
Uses an application defined "cache_id" to identify a connection.
This cache_id may be an IP address, DNS name or anything else the
client wishes to use.

Signed-off-by: Todd Short <todd.short@me.com>

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 216 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 247 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

vdukhovni commented on ssl/ssl_lib.c:
```diff
@@ -3836,12 +3877,26 @@ static unsigned long ssl_session_hash(const SSL_SESSION *a)
 static int ssl_session_cmp(const SSL_SESSION *a, const SSL_SESSION *b)
 {
+    /* if both have cache_id set, use that (assume client-side cache) */
+    if (a->cache_id != NULL && b->cache_id != NULL) {
```
What happens if one of `a` or `b` has a cache id and the other does not?

tmshort replied:
It should fail...

vdukhovni commented on ssl/ssl_local.h:
```diff
@@ -2959,6 +2964,8 @@ int ssl_srp_server_param_with_username_intern(SSL_CONNECTION *s, int *ad);
 
 void ssl_session_calculate_timeout(SSL_SESSION *ss);
 
+long ssl_ctx_set_sess_cache_mode(SSL_CTX *ctx, long new_mode);
```
Why is this needed?  We already have: `SSL_CTX_set_session_cache_mode(3)`.

tmshort replied:
It's not. It should be cleaned up.

vdukhovni commented on ssl/ssl_sess.c:
```diff
@@ -1457,4 +1471,193 @@ void SSL_CTX_set_stateless_cookie_verify_cb(
+    mode = sc->session_ctx->session_cache_mode;
+
+    /* Not client mode w/o server also */
+    if ((mode & SSL_SESS_CACHE_BOTH) != SSL_SESS_CACHE_CLIENT) {
+        ERR_raise(ERR_LIB_SSL, SSL_R_BAD_CACHE_MODE);
+        return NULL;
+    }
```
Why forbid modes that also have a server cache???

tmshort replied:
Hmmm... it "shouldn't" work, but actually, I think it will, because cache_ids are checed regardless during search. My original implementation used a separate function for searching, which is no longer the case.

vdukhovni commented on ssl/ssl_sess.c:
```diff
@@ -1457,4 +1471,193 @@ void SSL_CTX_set_stateless_cookie_verify_cb(
+        data.cache_id_len = sc->cache_id_len;
+        data.sid_ctx_length = sc->sid_ctx_length;
+        if (sc->sid_ctx_length > 0)
+            memcpy(data.sid_ctx, sc->sid_ctx, sc->sid_ctx_length);
```
Just as on the server a non-empty sid_ctx should be required I think.

tmshort replied:
`sid_ctx` is optional... On the server side, it checks to see if they are equal length, and then checks the value. The equal length can be 0. But in the case where SSL_SESS_CACHE_NO_INTERNAL_LOOKUP is examined, `sid_ctx` is not looked at.

tmshort replied:
Interestingly, when searching for a session (see `ssl_session_cmp`):

The server ignores `sid_ctx`, and only checks if they are equal later in `ssl_get_prev_session()`, where the `sid_ctx_length`s are checked, then the `sid_ctx` itself. Having an `sid_ctx_length` of 0 is valid (and subsequently considered equal).

The client checks to see if they have the same length (which could be 0), and the same value if the length is greater than 0.

vdukhovni commented on ssl/ssl_sess.c:
```diff
@@ -1457,4 +1471,193 @@ void SSL_CTX_set_stateless_cookie_verify_cb(
+
+        ret = sc->session_ctx->get_session_cb(s, sc->cache_id,
+                                             sc->cache_id_len,
+                                             &copy);
```
Should these also be checked for timeout, or is that solely up to the external cache?

tmshort replied:
It appears to be solely up to the external cached. It's not checked as a server.

vdukhovni commented on ssl/ssl_sess.c:
```diff
@@ -1457,4 +1471,193 @@ void SSL_CTX_set_stateless_cookie_verify_cb(
+    if (ss == NULL)
+        return 0;
+
+    if (!SSL_SESSION_set_timeout(ss, (long)t))
```
This mutates an extant session, I suspect this function should not be implemented. Any timeout should be set along with the cache id, before the start of the handshake. If the cache id is set first and automatically loads a session, the new timeout should be ignored, it is only for "fresh" sessions.

tmshort replied:
`SSL_SESSION_set_timeout()` is a pre-existing function.

tmshort replied:
But I will admit, that `SSL_SESSION_set_timeout_update_cache()` is orthogonal to this functionality.

vdukhovni commented on ssl/ssl_sess.c:
```diff
@@ -1457,4 +1471,193 @@ void SSL_CTX_set_stateless_cookie_verify_cb(
+    return 1;
+}
+
+int SSL_SESSION_set1_cache_id(SSL_SESSION *ss, const unsigned char *data, size_t len)
```
Likewise, I'm sceptical of the need for/safety of this function.

tmshort replied:
This returns an error if the SSL_SESSION is already in the cache.

vdukhovni replied:
@tmshort Let's chat about this interactively at some point before long. My timezone is GMT+1100, but will be GMT+1000 after April 6th.
Drop me an email with some proposed times...

tmshort replied:
@vdukhovni Not sure I have your email? But you should have mine.

As likely pointed out, this PR originated over 2 years ago, and has not been a priority for me after my job change... That said, I _do_ understand several of your concerns regarding modification of `SSL_SESSION`s that may be in the cache (and my attempt to protect them from modification), and making the client cache search mechanism automatic (like the server).

Since I can't remember the reason (or use-cases) for some of these decisions, I have been reconsidering them.
I've already reduced the scope of this by removing unrelated/unnecessary functions (they could be introduced later, but they are not needed for this functionality), so that should help a bit.

vdukhovni replied:
You can try `viktor@openssl.org`. I tried writing some time back about the RPK issue, but never heard back...

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

vdukhovni commented:
This PR has gone dormant. Closing.

tmshort commented on doc/man3/SSL_CTX_set_session_cache_mode.pod:
```diff
@@ -67,6 +69,7 @@ server will try to reuse the session.  This is the default.
 =item SSL_SESS_CACHE_BOTH
 
 Enable both SSL_SESS_CACHE_CLIENT and SSL_SESS_CACHE_SERVER at the same time.
+The search method uses session ids to retrieve sessions.
```
A given cache can only have a single search method; which ends up defaulting to the session ID, rather than the cache ID.

tmshort commented on doc/man3/SSL_CTX_set_session_cache_mode.pod:
```diff
@@ -49,12 +49,14 @@ No session caching for client or server takes place.
+This cache identifier must be set before calling L<SSL_connect(3)>.
+
+The application then can retrieve prior sessions via L<SSL_get1_previous_client_session(3)>
+and then assign the session to be reused via L<SSL_set_session(3)>.
```
It allows the user to choose to resume from the cache, or to create a new session, even if it means overwriting the session in cache.

tmshort replied:
For example, checking  `SSL_SESSION_get_verify_result(const SSL_SESSION *ss)`

tmshort replied:
This is mentioned in the SSL_set1_cache_id.pod documentation:

>After the session has been retrieved, it may be examined to determine whether
it can (or should) be used. The session must then be applied to the SSL object via the
L<SSL_set_session(3)> before it can be used during L<SSL_connect(3)>.

tmshort replied:
(Copied that text here.)

tmshort commented on doc/man3/SSL_SESSION_get_time.pod:
```diff
@@ -49,6 +52,10 @@ the long datatype instead of time_t and are therefore deprecated due to not
 being Y2038-safe on 32 bit systems. Note that such systems still need
 to be configured to use 64 bit time_t to be able to avoid overflow in system time.
 
+SSL_SESSION_set_timeout_update_cache() sets the timeout value for session
+referenced by B<ssl> in seconds to B<tm>, and ensures that any
+cached session that matches the SSL's session is updated as well.
```
It's in the code. It's an updated version of `SSL_SESSION_set_timeout()` that also updates the timeout of the session in the cache.
This is done via proper locking mechanism.

tmshort commented on doc/man3/SSL_set1_cache_id.pod:
```diff
@@ -0,0 +1,100 @@
+ int SSL_set1_cache_id(SSL *s, const unsigned char *data, size_t len);
+ int SSL_get0_cache_id(const SSL *s, const unsigned char **data, size_t *len);
+
+ int SSL_SESSION_set1_cache_id(SSL_SESSION *ss, const unsigned char *data, size_t len);
```
The code block modification of a session that's already in the cache.

tmshort commented on doc/man3/SSL_set1_cache_id.pod:
```diff
@@ -0,0 +1,100 @@
+The OpenSSL library can store/retrieve SSL/TLS sessions for later reuse.
+The sessions can be held in memory for each B<ctx>; if more than one
+SSL_CTX object is being maintained, the sessions are unique for each SSL_CTX
+object. The cache mode must be set to SSL_SESS_CACHE_CLIENT via
+L<SSL_CTX_set_session_cache_mode(3)> in order to use application-defined
+cache identifiers (otherwise, session IDs are used).
```
I'm not sure what you're asking for here? "the mode might enable oth" is unclear.

tmshort commented on doc/man3/SSL_set1_cache_id.pod:
```diff
@@ -0,0 +1,100 @@
+To reuse a session, use the SSL_set1_cache_id() function to specify the
+cache identifier of the server that the client is attempting to connect to. Set the
+optional session ID context via L<SSL_CTX_set_session_id_context(3)> or
+L<SSL_set_session_id_context(3)> before calling SSL_get1_previous_client_session().
```
Because the client may not want to use the session for who-know-why reasons. It's application-defined as to why this is the case, so applying a session is not forced onto the application.

tmshort commented on doc/man3/SSL_set1_cache_id.pod:
```diff
@@ -0,0 +1,100 @@
+
+After the session has been retrieved, it may be examined to determine whether
+it can (or should) be used. The session must then be applied to the SSL object via the
+L<SSL_set_session(3)> before it can be used during L<SSL_connect(3)>.
```
It depends on the application. Some connections may want to continue using a session, while other connections may need a fresh connection. There may be a need to have, e.g. a non-authenticated connection vs. an authenticated connection. The use case here is proxies, that may have to have different connections based on the source.

tmshort commented:
I plan to update this, based on comments.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 215 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 246 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 277 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 185 days ago

tmshort commented:
Updated, but still looking for review @vdukhovni

idrassi commented on ssl/ssl_lib.c:
```diff
@@ -2315,6 +2319,24 @@ int SSL_connect(SSL *s)
+     * If the cache is not set to client mode, an error will be raised, so
+     * pop off as necessary.
+     */
+    if (sc->session == NULL && sc->cache_id != NULL) {
```
I don't think it is safe to install a session from `cache_id` alone here. The new manual explicitly allows an IP address as the ID. If that IP serves more than one name, a session authenticated for one name can be resumed while this `SSL` is configured for another name. No new certificate check takes place during resumption, and the old verification result is carried into the connection.

I tested this with both TLS 1.2 and TLS 1.3. A full handshake for `other.example` failed with `X509_V_ERR_HOSTNAME_MISMATCH` as expected. With the same IP cache ID, OpenSSL resumed a session created for `server.example`: the connection succeeded and `SSL_get_verify_result` returned `X509_V_OK`.

The library needs to validate that the cached session matches the current peer identity and security context before installing it, or give the application a real chance to approve it. At present, even an application that calls `SSL_get1_previous_client_session`, examines the result, and rejects it will have the same session selected again by `SSL_connect`.

idrassi commented on ssl/ssl_lib.c:
```diff
@@ -4126,12 +4163,29 @@ static unsigned long ssl_session_hash(const SSL_SESSION *a)
 static int ssl_session_cmp(const SSL_SESSION *a, const SSL_SESSION *b)
 {
+    /* if both have cache_id set, use that (assume client-side cache) */
+    if (a->cache_id != NULL && b->cache_id != NULL) {
+        if (a->cache_id_len != b->cache_id_len
+            || memcmp(a->cache_id, b->cache_id, a->cache_id_len) != 0)
+            return 1;
+        if (a->sid_ctx_length != b->sid_ctx_length)
+            return 1;
+        if (a->sid_ctx_length > 0
+            && memcmp(a->sid_ctx, b->sid_ctx, a->sid_ctx_length) != 0)
+            return 1;
+        return 0;
+    }
```
This key can represent only one session for each cache ID and session ID context. That's not enough for TLS 1.3. A server may send several tickets, and each ticket should normally be used once. Here each new ticket replaces the previous one. The lookup later takes another reference instead of removing the ticket, so two handshakes starting at the same time can select the same ticket before the cleanup code removes it.

I confirmed this with a server issuing two tickets. The new session callback saw both tickets, the internal cache retained one entry, and two lookups returned the exact same `SSL_SESSION` pointer.

I think the client cache needs a collection of tickets for each cache ID, with an atomic take operation for TLS 1.3. TLS 1.2 can continue to use a reusable entry. Removing a ticket after a completed handshake doesn't close the race between concurrent handshakes.

idrassi reviewed (changes requested):
I spent some time testing these changes and I found two issues that I think need to be resolved at the design level.
